src/inference/ns.jl

```diff
@@ -0,0 +1,354 @@
+"""
+------------------------------------
+not working yet
+still being built
+refering ess.jl which is similar
+-------------------------------------
+
+Nested sampling algorithm.
+
+An example
+#Note-to-self: set a seed here
+## write an example here
+
+"""
+
+## what all to setup here:
+## 1. A subtype of AbstractSampler, defined as a mutable struct containing state information or sampler parameters... Partially done in  mutable struct NSState ? check
+## 2. A function sample_init! which performs any necessary set-up (default: do not perform any set-up)... Partially done in: function AbstractMCMC.sample_init! ? check
+## 3. A function step! which returns a transition that represents a single draw from the sampler... Partially done with: 2 function AbstractMCMC.step! ? check
+## 4. A function transitions_init which returns a container for the transitions obtained from the sampler (default: return a Vector{T} of length N where T is the type of the transition obtained in the first step and N is the number of requested samples).
+## 5. A function transitions_save! which saves transitions to the container (default: save the transition of iteration i at position i in the vector of transitions)
+## 6. A function sample_end! which handles any sampler wrap-up (default: do not perform any wrap-up)... Partially done in: function AbstractMCMC.sample_end! ? check
+## 7. A function bundle_samples which accepts the container of transitions and returns a collection of samples (default: return the vector of transitions)... Partially done with: 2 function AbstractMCMC.bundle_samples ? check
+
+## model:: NestedModel
+## sampler:: Nested
+
+## How the struct NestedModel is defined:
+##struct NestedModel <: AbstractModel
+##    loglike::Function
+##    prior_transform::Function
+##end
+
+##function NestedModel(loglike, priors::AbstractVector{<:UnivariateDistribution})
+##    prior_transform(X) = quantile.(priors, X)
+##    return NestedModel(loglike, prior_transform)
+##end
+
+## revisit the Nested sampler definition here: https://github.com/TuringLang/NestedSamplers.jl/blob/master/src/staticsampler.jl
+
+## comment on each step of the code and also include docstrings for sections of the code
+## both NestedModel and Nested are to be utilized in this code
+
+struct NS{space, T, P <: AbstractProposal, B <: AbstractBoundingSpace{T}} <: InferenceAlgorithm  ## refer the comment on line 39
```

Comment:
No need to duplicate all these fields, just make `NS` wrap around `Nested` from [here](https://github.com/TuringLang/NestedSamplers.jl/blob/19f5273b25a5eb3a6b1278f98a1dacf5b302acb0/src/staticsampler.jl#L3).

Reply:
See comment below -- just make a `Nested` struct and stick it as a field in `NSState`. Don't duplicate all this code. Rely on `NestedSamplers.jl` code as much as possible.